### PR TITLE
Change priority function(CalculateAntiAffinityPriority) into Map/Reduce pattern

### PR DESCRIPTION
**Pull request by guangxuli**
**What this PR does / why we need it**:
Ref #24246. exactly ref https://github.com/kubernetes/kubernetes/issues/51455, the PR aim to unify priority functions(deprecated) by using map/reduce pattern. 
Previous related PR is https://github.com/kubernetes/kubernetes/pull/51192
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
None
**Special notes for your reviewer**:

**Release note**:

```release-note
None
```

### CODE REVIEW

**Comment by xiangpengzhao** [comment]
/ok-to-test

cc @kubernetes/sig-scheduling-pr-reviews

**Comment by guangxuli** [comment]
@bsalamat @wojtek-t @resouer @aveshagarwal

**Comment by wojtek-t** [comment]
This is too risky to get into 1.9 - I will take a look closer to the end of code freeze.

**Comment by guangxuli** [comment]
@wojtek-t sure, thanks.

**Comment by guangxuli** [comment]
@wojtek-t Do you have time slot to review this?

**Comment by wojtek-t** [comment]
will get to this soon

**Comment by guangxuli** [comment]
@wojtek-t Thanks a lot. :)

File: plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go [hunk 1]
```diff
@@ -201,52 +201,76 @@ func (s *ServiceAntiAffinity) getNodeClassificationByLabels(nodes []*v1.Node) (m
...
-	numServicePods := len(nsServicePods)
-	result := []schedulerapi.HostPriority{}
+
 	//score int - scale of 0-maxPriority
 	// 0 being the lowest priority and maxPriority being the highest
-	for node := range labeledNodes {
+	for i, hostPriority := range result {
+		label, ok := labelNodesStatus[hostPriority.Host]
+		if !ok {
+			result[i] = schedulerapi.HostPriority{Host: hostPriority.Host, Score: int(0)}
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go]: result[i].Score = 0 (or int(0))

File: plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go [hunk 2]
```diff
@@ -201,52 +201,76 @@ func (s *ServiceAntiAffinity) getNodeClassificationByLabels(nodes []*v1.Node) (m
...
 		if numServicePods > 0 {
-			fScore = float64(schedulerapi.MaxPriority) * (float64(numServicePods-podCounts[labeledNodes[node]]) / float64(numServicePods))
+			fScore = maxPriorityFloat64 * (float64(numServicePods-podCounts[label]) / float64(numServicePods))
 		}
-		result = append(result, schedulerapi.HostPriority{Host: node, Score: int(fScore)})
-	}
-	// add the open nodes with a score of 0
-	for _, node := range nonLabeledNodes {
-		result = append(result, schedulerapi.HostPriority{Host: node, Score: 0})
+		result[i] = schedulerapi.HostPriority{Host: hostPriority.Host, Score: int(fScore)}
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go]: result[i].Score = int(fScore)

**Review by wojtek-t — COMMENTED** [review]
Generally looks ok. Just a couple comments.

File: plugin/pkg/scheduler/algorithm/priorities/metadata.go [hunk 3]
```diff
@@ -44,11 +44,12 @@ func NewPriorityMetadataFactory(serviceLister algorithm.ServiceLister, controlle
...
-	podTolerations []v1.Toleration
-	affinity       *v1.Affinity
-	podSelectors   []labels.Selector
-	controllerRef  *metav1.OwnerReference
+	nonZeroRequest         *schedulercache.Resource
+	podTolerations         []v1.Toleration
+	affinity               *v1.Affinity
+	podSelectors           []labels.Selector
+	controllerRef          *metav1.OwnerReference
+	podSelectorsOneService labels.Selector
```
[the remark this answers is not in the document]
Comment (guangxuli) [reply plugin/pkg/scheduler/algorithm/priorities/metadata.go]: agree,will do that.

File: plugin/pkg/scheduler/algorithm/priorities/metadata.go [hunk 4]
```diff
@@ -59,41 +60,64 @@ func (pmf *PriorityMetadataFactory) PriorityMetadata(pod *v1.Pod, nodeNameToInfo
 	}
 	tolerationsPreferNoSchedule := getAllTolerationPreferNoSchedule(pod.Spec.Tolerations)
 	podSelectors := getSelectors(pod, pmf.serviceLister, pmf.controllerLister, pmf.replicaSetLister, pmf.statefulSetLister)
+	firstServiceSelector := getOneServiceSelector(pod, pmf.serviceLister)
```
[the remark this answers is not in the document]
Comment (guangxuli) [reply plugin/pkg/scheduler/algorithm/priorities/metadata.go]: agree

File: plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go [hunk 5]
```diff
@@ -201,52 +201,76 @@ func (s *ServiceAntiAffinity) getNodeClassificationByLabels(nodes []*v1.Node) (m
...
-			return nil, err
-		}
-		// consider only the pods that belong to the same namespace
-		for _, nsPod := range pods {
-			if nsPod.Namespace == pod.Namespace {
-				nsServicePods = append(nsServicePods, nsPod)
-			}
+// filteredPod get pods based on namespace and selector
+func filteredPod(nameSpace string, selector labels.Selector, nodeInfo *schedulercache.NodeInfo) (pods []*v1.Pod) {
+	if nodeInfo.Pods() == nil || len(nodeInfo.Pods()) == 0 || selector == nil {
```
[the remark this answers is not in the document]
Comment (guangxuli) [reply plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go]: because selector may be nil in current unit tests, it will result panic if not add selector validation.

File: plugin/pkg/scheduler/algorithm/priorities/metadata.go [hunk 6]
```diff
@@ -59,41 +60,64 @@ func (pmf *PriorityMetadataFactory) PriorityMetadata(pod *v1.Pod, nodeNameToInfo
...
+		if services, err := sl.GetPodServices(pod); err == nil && len(services) > 0 {
+			return labels.SelectorFromSet(services[0].Spec.Selector)
+		}
+	}
+	return nil
+}
+
 // getSelectors returns selectors of services, RCs and RSs matching the given pod.
-func getSelectors(pod *v1.Pod, sl algorithm.ServiceLister, cl algorithm.ControllerLister, rsl algorithm.ReplicaSetLister, ssl algorithm.StatefulSetLister) []labels.Selector {
+func getSelectors(pod *v1.Pod, sl algorithm.ServiceLister, cl algorithm.ControllerLister, rsl algorithm.ReplicaSetLister, ssl algorithm.StatefulSetLister) (all []labels.Selector) {
```
[the remark this answers is not in the document]
Comment (guangxuli) [reply plugin/pkg/scheduler/algorithm/priorities/metadata.go]: done

File: plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go [hunk 7]
```diff
@@ -201,52 +201,76 @@ func (s *ServiceAntiAffinity) getNodeClassificationByLabels(nodes []*v1.Node) (m
...
-		if err != nil {
-			return nil, err
-		}
-		// consider only the pods that belong to the same namespace
-		for _, nsPod := range pods {
-			if nsPod.Namespace == pod.Namespace {
-				nsServicePods = append(nsServicePods, nsPod)
-			}
+// filteredPod get pods based on namespace and selector
+func filteredPod(nameSpace string, selector labels.Selector, nodeInfo *schedulercache.NodeInfo) (pods []*v1.Pod) {
```
[the remark this answers is not in the document]
Comment (guangxuli) [reply plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go]: done

File: plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go (hunk 1 above)
Comment (guangxuli) [reply plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go]: done

File: plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go (hunk 2 above)
Comment (guangxuli) [reply plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go]: done

**Comment by guangxuli** [comment]
thanks @wojtek-t  all comments addressed.

File: plugin/pkg/scheduler/algorithm/priorities/metadata.go [hunk 8]
```diff
@@ -57,43 +58,63 @@ func (pmf *PriorityMetadataFactory) PriorityMetadata(pod *v1.Pod, nodeNameToInfo
...
+		podTolerations:          getAllTolerationPreferNoSchedule(pod.Spec.Tolerations),
+		affinity:                pod.Spec.Affinity,
+		podSelectors:            getSelectors(pod, pmf.serviceLister, pmf.controllerLister, pmf.replicaSetLister, pmf.statefulSetLister),
+		controllerRef:           priorityutil.GetControllerRef(pod),
+		podFirstServiceSelector: getOneServiceSelector(pod, pmf.serviceLister),
 	}
 }
 
+// getOneServiceSelector returns one selector of services the given pod.
+func getOneServiceSelector(pod *v1.Pod, sl algorithm.ServiceLister) (firstServiceSelector labels.Selector) {
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/metadata.go]: getFirstServiceSelector maybe (for consistency)?

File: plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go [hunk 9]
```diff
@@ -201,52 +201,79 @@ func (s *ServiceAntiAffinity) getNodeClassificationByLabels(nodes []*v1.Node) (m
...
-			return nil, err
-		}
-		// consider only the pods that belong to the same namespace
-		for _, nsPod := range pods {
-			if nsPod.Namespace == pod.Namespace {
-				nsServicePods = append(nsServicePods, nsPod)
-			}
+// filteredPod get pods based on namespace and selector
+func filteredPod(namespace string, selector labels.Selector, nodeInfo *schedulercache.NodeInfo) (pods []*v1.Pod) {
+	if nodeInfo.Pods() == nil || len(nodeInfo.Pods()) == 0 || selector == nil {
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go]: Is this "selector ==nil" added only for the purpose of tests?
If so, I'm against adding it and we should change tests to always pass selector.

**Comment by wojtek-t** [comment]
Also - please squash the commits.

**Comment by guangxuli** [comment]
@wojtek-t yes, done.

**Comment by wojtek-t** [comment]
See also my comments in the PR itself - those aren't addressed.

**Comment by guangxuli** [comment]
@wojtek-t  oh, sorry, didn't see that, will fix.

File: plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go (hunk 9 above)
Comment (guangxuli) [reply plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go]: In TestZoneSpreadPriority  there are special test case - e.g. no service/nothing scheduled/different services  test cases. In these special test cases "selector == nil" is reasonable. WDYT?

File: plugin/pkg/scheduler/algorithm/priorities/metadata.go [hunk 10]
```diff
@@ -57,43 +58,63 @@ func (pmf *PriorityMetadataFactory) PriorityMetadata(pod *v1.Pod, nodeNameToInfo
...
+		affinity:                pod.Spec.Affinity,
+		podSelectors:            getSelectors(pod, pmf.serviceLister, pmf.controllerLister, pmf.replicaSetLister, pmf.statefulSetLister),
+		controllerRef:           priorityutil.GetControllerRef(pod),
+		podFirstServiceSelector: getFirstServiceSelector(pod, pmf.serviceLister),
 	}
 }
 
+// getFirstServiceSelector returns one selector of services the given pod.
+func getFirstServiceSelector(pod *v1.Pod, sl algorithm.ServiceLister) (firstServiceSelector labels.Selector) {
+	if sl != nil {
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/metadata.go]: Can we get rid of this if (and assume it's non-nil)? If this is nil, this is a bug in code and should be fixed.
[See also my comment below]

File: plugin/pkg/scheduler/algorithm/priorities/metadata.go [hunk 11]
```diff
@@ -57,43 +58,63 @@ func (pmf *PriorityMetadataFactory) PriorityMetadata(pod *v1.Pod, nodeNameToInfo
...
+}
+
 // getSelectors returns selectors of services, RCs and RSs matching the given pod.
 func getSelectors(pod *v1.Pod, sl algorithm.ServiceLister, cl algorithm.ControllerLister, rsl algorithm.ReplicaSetLister, ssl algorithm.StatefulSetLister) []labels.Selector {
 	var selectors []labels.Selector
-	if services, err := sl.GetPodServices(pod); err == nil {
-		for _, service := range services {
-			selectors = append(selectors, labels.SelectorFromSet(service.Spec.Selector))
+
+	if sl != nil {
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/metadata.go]: Hmm - I'm against introducing those "ifs" here.
If those listers are not passed, this is a bug in code and should be fixed. With those ifs, we are risking not observing this.

Comment (guangxuli) [reply plugin/pkg/scheduler/algorithm/priorities/metadata.go]: Actually, these "ifs" are also avoid of panic in test case "TestZoneSpreadPriority". In this test case rsl/ssl/cl are all set to nil expect service. if we set a specific non-useful rsl/ssl/cl to avoid this panic in this test case, looks like redundant is some degree.

Comment (wojtek-t) [reply plugin/pkg/scheduler/algorithm/priorities/metadata.go]: Let's not add production code to fix test issues. Change test in that case.

Comment (wojtek-t) [reply plugin/pkg/scheduler/algorithm/priorities/metadata.go]: And if the test would be the same as another one, let's remove it.

Comment (guangxuli) [reply plugin/pkg/scheduler/algorithm/priorities/metadata.go]: I have added three local variables (sss/rcs/rss)  in TestZoneSpreadPriority to make sure the parameter not nil when construct metaDataProducer. and remove the  related "ifs".

File: plugin/pkg/scheduler/algorithm/priorities/metadata.go [hunk 12]
```diff
@@ -57,43 +58,56 @@ func (pmf *PriorityMetadataFactory) PriorityMetadata(pod *v1.Pod, nodeNameToInfo
...
+
 // getSelectors returns selectors of services, RCs and RSs matching the given pod.
 func getSelectors(pod *v1.Pod, sl algorithm.ServiceLister, cl algorithm.ControllerLister, rsl algorithm.ReplicaSetLister, ssl algorithm.StatefulSetLister) []labels.Selector {
 	var selectors []labels.Selector
+
 	if services, err := sl.GetPodServices(pod); err == nil {
 		for _, service := range services {
-			selectors = append(selectors, labels.SelectorFromSet(service.Spec.Selector))
+			selector := labels.SelectorFromSet(service.Spec.Selector)
+			selectors = append(selectors, selector)
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/metadata.go]: Why did you change that? We are now unnecessary allocating one variable.

**Review by wojtek-t — COMMENTED** [review]
Just last minor nit.

File: plugin/pkg/scheduler/algorithm/priorities/metadata.go (hunk 12 above)
Comment (guangxuli) [reply plugin/pkg/scheduler/algorithm/priorities/metadata.go]: agree, that is unnecessary, will fix.
